From db0d060042a183b52c7ea5c477e059c3a74d222a Mon Sep 17 00:00:00 2001
From: Rathees Kumar R Chinannan <quic_rrchinan@quicinc.com>
Date: Tue, 10 Sep 2024 14:37:02 +0530
Subject: [PATCH 4/4] hostapd: Add configuration for Multi MBSSID group mode

Add configuration parameter to set group size on Multi MBSSID
group mode.

 0 = Disabled
 1 = Multiple BSSID advertisement enabled.
 2 = Enhanced multiple BSSID advertisement enabled.
 3 = Multi MBSSID group enabled

Patch-dependency: q01-001-mbssid-add-configuration-options.patch

Signed-off-by: Rathees Kumar R Chinannan <quic_rrchinan@quicinc.com>
---
 hostapd/config_file.c | 24 +++++++++++++++++++++++-
 hostapd/hostapd.conf  |  7 +++++++
 src/ap/ap_config.c    |  1 +
 src/ap/ap_config.h    |  4 ++++
 4 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 873abde..fe14652 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -3876,13 +3876,35 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		conf->discard_6g_awgn_event = atoi(pos);
 	} else if (os_strcmp(buf, "mbssid") == 0) {
 		int mbssid = atoi(pos);
-		if (mbssid < 0 || mbssid > ENHANCED_MBSSID_ENABLED) {
+		if (mbssid < 0 || mbssid > MULTI_MBSSID_GROUP_ENABLED) {
 			wpa_printf(MSG_ERROR,
 				   "Line %d: invalid mbssid (%d): '%s'.",
 				   line, mbssid, pos);
 			return 1;
 		}
 		conf->mbssid = mbssid;
+	} else if (os_strcmp(buf, "mbssid_group_size") == 0) {
+		int group_size = atoi(pos);
+		int val = group_size;
+
+		if (group_size < MULTI_MBSSID_GROUP_SIZE_MIN ||
+		    group_size > MULTI_MBSSID_GROUP_SIZE_MAX) {
+			wpa_printf(MSG_ERROR,
+				   "Line %d: invalid group_size, allowed range [%d to %d]",
+				   line, MULTI_MBSSID_GROUP_SIZE_MIN,
+				   MULTI_MBSSID_GROUP_SIZE_MAX);
+			return 1;
+		}
+		while (val != 1) {
+			if (val % 2 != 0) {
+				wpa_printf(MSG_ERROR,
+					   "Line %d: invalid group_size, not power of 2",
+					   line);
+				return 1;
+			}
+			val = val / 2;
+		}
+		conf->group_size = group_size;
 #endif /* CONFIG_IEEE80211AX */
 	} else if (os_strcmp(buf, "max_listen_interval") == 0) {
 		bss->max_listen_interval = atoi(pos);
diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index 54ff4a6..96637ba 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -3366,6 +3366,7 @@ own_ip_addr=127.0.0.1
 # 0 = Disabled (default)
 # 1 = Multiple BSSID advertisement enabled.
 # 2 = Enhanced multiple BSSID advertisement enabled.
+# 3 = Multi MBSSID group enabled
 #mbssid=0
 #
 # The transmitting interface should be added with the 'interface' option while
@@ -3399,3 +3400,9 @@ own_ip_addr=127.0.0.1
 #bridge=br-lan
 #wpa_key_mgmt=SAE
 #bssid=00:03:7f:12:84:85
+#
+# Multi MBSSID group size
+# This field is used to configure the number of BSSes in each group
+# when Multi MBSSID group support is enabled for the radio.
+# This value should be in power of 2.
+#mbssid_group_size=4
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index e469d73..f222d32 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -334,6 +334,7 @@ struct hostapd_config * hostapd_config_defaults(void)
 #ifdef CONFIG_AIRTIME_POLICY
 	conf->airtime_update_interval = AIRTIME_DEFAULT_UPDATE_INTERVAL;
 #endif /* CONFIG_AIRTIME_POLICY */
+	conf->group_size = MULTI_MBSSID_GROUP_SIZE_DEFAULT;
 
 	hostapd_set_and_check_bw320_offset(conf, 0);
 
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 78f6265..ccc3e0d 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1273,6 +1273,10 @@ struct hostapd_config {
 	int ccfs;
 	int bandwidth_device;
 	int center_freq_device;
+
+#define MULTI_MBSSID_GROUP_SIZE_MIN 2
+#define MULTI_MBSSID_GROUP_SIZE_MAX 16
+#define MULTI_MBSSID_GROUP_SIZE_DEFAULT 4
 	u8 group_size;
 };
 
-- 
2.34.1

